File: veterinaria-system/backend/api/controllers/tutor_controller.py
"""
Controller para el modelo Tutor.
Arquitectura MVC - Capa de Controladores
"""
from api.models import Tutor


class TutorController:
    """
    Controlador para la lógica de negocio de Tutores.
    """

    @staticmethod
    def get_all_tutors():
        """Obtiene todos los tutores"""
        return Tutor.objects.select_related('usuario').all()

    @staticmethod
    def get_tutor_by_id(tutor_id):
        """Obtiene un tutor por ID"""
        return Tutor.objects.select_related('usuario').get(id=tutor_id)

    @staticmethod
    def get_tutor_by_ci(ci):
        """Obtiene un tutor por cédula de identidad"""
        return Tutor.objects.filter(ci=ci).select_related('usuario').first()

    @staticmethod
    def get_tutor_pets(tutor_id):
        """Obtiene todas las mascotas de un tutor"""
        tutor = Tutor.objects.get(id=tutor_id)
        return tutor.mascotas.filter(activo=True)

    @staticmethod
    def create_tutor(data):
        """Crea un nuevo tutor"""
        return Tutor.objects.create(**data)

    @staticmethod
    def update_tutor(tutor_id, data):
        """Actualiza un tutor existente"""
        tutor = Tutor.objects.get(id=tutor_id)
        for key, value in data.items():
            setattr(tutor, key, value)
        tutor.save()
        return tutor

    @staticmethod
    def delete_tutor(tutor_id):
        """Elimina un tutor"""
        tutor = Tutor.objects.get(id=tutor_id)
        tutor.delete()

File: veterinaria-system/frontend/src/controllers/TutorController.js
/**
 * Controller para Tutores
 * Arquitectura MVC - Capa de Controladores
 */

import api from '../services/api';
import { Tutor, Pet } from '../models';

class TutorController {
  /**
   * Obtiene todos los tutores
   * @returns {Promise<Tutor[]>} Lista de tutores
   */
  static async getAll() {
    try {
      const response = await api.get('/tutores/');
      return response.data.map(tutor => Tutor.fromJSON(tutor));
    } catch (error) {
      throw new Error('Error al obtener tutores');
    }
  }

  /**
   * Obtiene un tutor por ID
   * @param {number} id - ID del tutor
   * @returns {Promise<Tutor>} Tutor
   */
  static async getById(id) {
    try {
      const response = await api.get(`/tutores/${id}/`);
      return Tutor.fromJSON(response.data);
    } catch (error) {
      throw new Error('Error al obtener tutor');
    }
  }

  /**
   * Obtiene las mascotas de un tutor
   * @param {number} tutorId - ID del tutor
   * @returns {Promise<Pet[]>} Lista de mascotas
   */
  static async getPets(tutorId) {
    try {
      const response = await api.get(`/tutores/${tutorId}/mascotas/`);
      return response.data.map(pet => Pet.fromJSON(pet));
    } catch (error) {
      throw new Error('Error al obtener mascotas del tutor');
    }
  }

  /**
   * Crea un nuevo tutor
   * @param {Object} tutorData - Datos del tutor
   * @returns {Promise<Tutor>} Tutor creado
   */
  static async create(tutorData) {
    try {
      const response = await api.post('/tutores/', tutorData);
      return Tutor.fromJSON(response.data);
    } catch (error) {
      throw new Error('Error al crear tutor');
    }
  }

  /**
   * Actualiza un tutor
   * @param {number} id - ID del tutor
   * @param {Object} tutorData - Datos actualizados
   * @returns {Promise<Tutor>} Tutor actualizado
   */
  static async update(id, tutorData) {
    try {
      const response = await api.put(`/tutores/${id}/`, tutorData);
      return Tutor.fromJSON(response.data);
    } catch (error) {
      throw new Error('Error al actualizar tutor');
    }
  }

  /**
   * Elimina un tutor
   * @param {number} id - ID del tutor
   */
  static async delete(id) {
    try {
      await api.delete(`/tutores/${id}/`);
    } catch (error) {
      throw new Error('Error al eliminar tutor');
    }
  }
}

export default TutorController;
